libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,387 @@
+/*
+ * AVL Tree implementation.
+ * Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+ */
+
+#ifndef UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+#define UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+
+#include <cstdlib>
+#include <cassert>
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <uavcan/build_config.hpp>
+#include <uavcan/util/templates.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/debug.hpp>
+
+namespace uavcan 
+{
+template <typename T>
+class UAVCAN_EXPORT AvlTree : Noncopyable 
+{
+protected:
+    struct Node {
+        T* data = UAVCAN_NULLPTR;
+        int16_t h = 1; // initially added as leaf
```

Comment:
MISRA prohibits native types because of associated portability issues and risks. The rest of the codebase is mostly non-compliant in this sense, but as I wrote in #190, we'll have a chance to have those non-compliance issues rectified in BlueSky. If `int16_t` is considered too small (which I don't think is the case here considering the purpose of this container), it should be replaced with `int32_t`.

Reply:
I'm not sure you are reading MISRA correctly here. Rule 3-9-2 (which is only Advisory) recommends typedefs to manually control the size of integer types. The intent of all the integer rules is to prevent defects. My recommendation to use a machine word is also to prevent alignment defects (which are especially insidious where un-aligned access is supported but slow). If we must we can use a ptrdiff_t but that'd just look weird. Let's feel free to add "breaks MIRSRA x-x-x because foo is safer" comments (we can even make them Coverity annotations) rather than slavishly adhering to the standard. But I do agree we should consider and use MISRA where it make sense.

That all said, I don't think we're packing these nodes into anything so my comment here is probably just cosmetic.

Reply:
Note: it can't be unsigned because it may be negative while performing balancing.

Reply:
yep. ptrdif_t and int are both signed machine words on all platforms...except 8-bit platforms...crap. int is 16 bits on AVR according to the spec (caveat: most AVR program make use of typedef "fast int" which is 8-bits) and ptrdiff_t can be larger than a machine word. Okay. I'm really over-indexing on this, sorry. Let's keep the int16_t and I'll get to the part where this data structure is actually used to be sure we're not causing un-aligned accesses on some platforms.

Reply:
> I'm not sure you are reading MISRA correctly here. Rule 3-9-2 (which is only Advisory) recommends typedefs to manually control the size of integer types.

Indeed, I was misremembering things by thinking that the rule is required rather than advisory in MISRA. Still, well-defined, deterministic width may eliminate certain overflow-related defects. One example comes to mind -- there was a discussion (either on the mailing list or on Gitter) about this code misbehaving on AVR due to integer overflow: https://github.com/UAVCAN/libcanard/pull/17/files#diff-ec9999a8cb27a176bf926dd68dfd7426R303

I have completely eliminated native types from libcanard since then; the only thing that stopped me from doing the same in libuavcan is the size of its codebase.

HIC++ also agrees here:

> To enhance portability, instead of using the standard integer types (signed char, short, int, long, long long, and the unsigned counterparts), size specific types should be defined in a project-wide header file, so that the definition can be updated to match a particular platform (16, 32 or 64bit). Where available, intN t and uintN t types (e.g. int8 t) defined in the cstdint header file should be used for this purpose.


I don't see how the alignment argument is relevant at all, considering that we don't use packed structs ever.  Consequently, I don't see how reliance on native types or platform-defined types like `ptrdiff_t` makes the code less prone to defects.

Generally, we should emphasize robustness over optimization.